### ISSUE TITLE
Is there a special 'hack' for operators and string literal types?

### ISSUE THREAD

**Issue by zhuravlikjb** [Microsoft/TypeScript#6505]
```typescript
type WeekDay = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday";

function doFun(a: WeekDay | WeekDay[]) {
    a = "Monday";  // string literal is contextually typed, so it has string literal type "Monday" => no type mismatch
    if (a === "Monday") { // string literal is NOT contextually typed and has type 'string' => should be type mismatch error
        
    }
}
```

Type mismatch error in the second case should be because:
1) we're comparing 'string' with "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday" | ("Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday")[] for assignability in both directions
2) according to the spec, in direction 'string' => the-other-type, we're comparing 'string' with all union type constituents, and it's not assignable to any of them => fail
3) according to the spec, in direction the-other-type => 'string', all constituents of the-other-type should be assignable to 'string', but obviously Array is not assignable to it => fail
4) both conditions fail => type mismatch error for '===' operator

But by some reason the compiler doesn't show any errors here.

Is there some special handling of string literal types with operators like '===', switch labels, etc?

**Comment by zhuravlikjb** [comment]
Ah, I've found it, sorry:
https://github.com/Microsoft/TypeScript/commit/8dbfe1ca631111e28a59b62e1ddc941ea6fe2e3f

**Comment by DanielRosenwasser** [comment]
We have two alternative proposals to solving this. One is inferring string literal types at comparison locations (see the PR at #6196), and the other is described on https://github.com/Microsoft/TypeScript/issues/6167#issuecomment-171455369.

**Comment by zhuravlikjb** [comment]
Thanks a lot, Daniel! Will track both PRs to see the final decision taken.

**Comment by DanielRosenwasser** [comment]
@zhuravlikjb things are leaning towards literal locations.
